### DISCUSSION TITLE
SOREL CAN BUS Interface (SCBI)

### DISCUSSION THREAD

**Discussion by javier6d2** [arendst/Tasmota#20598] [Support request] [answered]
Sorel is a manufacturer of difference controllers for heating and solar systems that use CAN Bus protocol

It would be possible to adapt this driver (xsns_87_mcp2515.ino) so the Sorela library showed bellow could be used to translate Sorel CAN Bus messages?

[Sorela library](https://github.com/aprospero/cansorella)

Kind regards

**Comment by gemu2015** [comment]
the sml driver supports CAN reading and writing on an ESP32 native can interface. there is an example in the docs
forget about the mcp2515, it is simply too slow.

**Comment by javier6d2** [comment]
So, I have built a tasmota32 firmware with smart meter interface feature and made the following set-up:

ESP32 + CAN transceiver + Sorel controller

![WhatsApp Image 2024-01-27 at 18 33 36](https://github.com/arendst/Tasmota/assets/16863195/150b1108-b7aa-47da-962a-de9bc0a3de58)

Configuration module tasmota is: GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)

Scripting to get the values from the solar controller overwhelmed me, should be something like:

`>D

>B
=>sensor53 r
>M 1
; Sorel
+1,15,C,0,3203,Sorel

#`

I found also this [script](https://github.com/ddimension/can-ltdc-logger/blob/master/can-ltdc-logger.c) for the same solar controller. May be it help to get the reads. I appreciate any light.

**Reply by gemu2015** [reply]
You need only a can transceiver, since the esp32 handles the canbus internally
https://www.ebay.de/itm/263690841938?chn=ps&_ul=DE&norover=1&mkevt=1&mkrid=707-134425-41852-0&mkcid=2&mkscid=101&itemid=263690841938&targetid=1404115578853&device=c&mktype=pla&googleloc=9044703&poi=&campaignid=17935704717&mkgroupid=139162549385&rlsatarget=pla-1404115578853&abcId=9301059&merchantid=138392580&gclid=EAIaIQobChMIxZn1263-gwMVVZtoCR2fgQ47EAQYAiABEgKenPD_BwE
you then specify the receive and transmit pins in sml descriptor
I will have a look at the script you provided and give some hints when you have the hardware ready

**Comment by javier6d2** [comment]
Thank you. I have already purchased the item. I'll come back when I have the correct hardware setup.

**Comment by javier6d2** [comment]
Hi gemu,

I have already received the can transceiver, I have tried the following script and executed in the console: sensor53 d1, without success. I don´t receive any telegram in the console.

```
>D
>B
=>sensor53 r
>M 1
; SOREL LTDC
; params -> 04 = baudrate 250kb + number of receive buffers * 100
+1,3,C,0,3204,CAN,1,5
1,1081407f0801800000UUuuUUuu@1024,Temp S1,C,tmpS1,1
; I don´t know which is the correct addresse
#
```

**Comment by gemu2015** [comment]
you must compile with additional option
#define USE_SML_CANBUS
if your Sorel controller emits telegrams without request you should see them in dump mode

**Comment by javier6d2** [comment]
Thank you. Now I can read the telegrams  in the console :

```
20:04:38.769 MQT: stat/tasmota_4DCF00/RESULT = {"Time":"2024-02-19T20:04:38","SML":{"CMD":"dump: 1"}}
20:05:23.183 : 01042480: 080900000000000000
20:05:23.185 : 10012480: 05000180ffff
20:05:23.186 : 09042480: 08005265696ec29c63
20:05:23.186 : 10032480: 025331
20:05:23.187 : 1008000b: 0824007c0000000018
20:05:23.199 : 10012480: 05000180ffff
20:05:23.200 : 10082480: 030a0500
20:05:23.200 : 09042480: 03fe696f
```

**Comment by javier6d2** [comment]
How I can define the Meter Metrics: `<M>,<decoder>@<scale><offs>,<label>,<UoM>,<var>,<precision>` to get the values in the console?. I have the following sensor temperature value in Sorel controler: S1=123

I get the following dump in the console:
```
16:40:00.966 : 00012485: 00
16:40:07.657 : 10012480: 0500ce040000
16:40:10.964 : 10012485: 0101
16:40:10.966 : 00012485: 00
```
I think that line: `16:40:07.657 : 10012480: 0500ce040000` means (HEX to decimal) ce = 206, 04=4 then 206 + 4*256=1233, then S1=1233/10=123ºC

I apreciate any help

**Comment by javier6d2** [comment]
`16:48:18.100 : 10012480: 050208020000` means (HEX to decimal) 08 = 8, 02=2 then 08 + 2x256= 520, then S3=520/10=52ºC
`16:52:50.301 : 10012480: 05034e020000` means (HEX to decimal) 4e = 78, 02=2 then 78+ 2x256= 590, then S4=590/10=59ºC

**Reply by gemu2015** [reply, marked as answer]
normally the specs of your meter should say how the telegrams are composed:
 10012480 is the ID of the telegram, the 05 is length of payload.
now you have 5 bytes payload:

in case your payload is a a word with 0802  (08020000)
you define a decoder with e.g unsigned word (defined with UUuu) 

1,100124800500uuUU@10,Temp S1,C,tmpS1,1

see the docs about the decoder options 
you should know the byte order and format of your payload

10 is the scaling factor from your sorel docs

**Reply by javier6d2** [reply]
Many many thanks the explanation and to point me in the right direction. I will continue watching the telegrams to extract more information from the Sorel controller. At this moment I have the temp values:

```
1,100124800500uuUU@10,Temp S1,ºC,S1,1
1,100124800501uuUU@10,Temp S2,ºC,S2,1
1,100124800502uuUU@10,Temp S3,ºC,S3,1
1,100124800503uuUU@10,Temp S4,ºC,S4,1 
```

**Reply by arnwik** [reply]
@gemu2015 
You mentioned that 10 is the scaling factor from the sorel docs. Where did you find that documentation?
Mine look something like this, and I have no clue which is sensor 1,2,3 and which is the relay for the pump etc:

Timestamp: 1729976452.338099    ID: 0001d085    X Rx                DL:  0                                Channel: can0
Timestamp: 1729976453.214224    ID: 1007d080    X Rx                DL:  8    25 00 01 00 00 00 00 00     Channel: can0
Timestamp: 1729976453.314203    ID: 1007d080    X Rx                DL:  8    25 01 00 00 00 00 00 00     Channel: can0
Timestamp: 1729976453.414220    ID: 1007d080    X Rx                DL:  8    25 02 01 00 00 00 00 00     Channel: can0
Timestamp: 1729976453.814187    ID: 1007d080    X Rx                DL:  8    40 00 01 00 00 00 00 00     Channel: can0
Timestamp: 1729976453.914211    ID: 1007d080    X Rx                DL:  8    40 01 00 00 00 00 00 00     Channel: can0
Timestamp: 1729976454.014234    ID: 1007d080    X Rx                DL:  8    40 02 01 00 00 00 00 00     Channel: can0
Timestamp: 1729976456.214277    ID: 1007d080    X Rx                DL:  8    60 00 11 00 00 00 00 00     Channel: can0
Timestamp: 1729976456.314253    ID: 1007d080    X Rx                DL:  8    60 01 00 00 00 00 00 00     Channel: can0
Timestamp: 1729976456.414254    ID: 1007d080    X Rx                DL:  8    60 02 11 00 00 00 00 00     Channel: can0
Timestamp: 1729976459.814328    ID: 1007d080    X Rx                DL:  8    80 00 b6 02 00 00 00 00     Channel: can0
Timestamp: 1729976459.914350    ID: 1007d080    X Rx                DL:  8    80 01 00 00 00 00 00 00     Channel: can0
Timestamp: 1729976460.014339    ID: 1007d080    X Rx                DL:  8    80 02 b6 02 00 00 00 00     Channel: can0
Timestamp: 1729976461.914407    ID: 1007d080    X Rx                DL:  8    c0 00 00 00 00 00 00 00     Channel: can0
Timestamp: 1729976462.014358    ID: 1007d080    X Rx                DL:  8    e0 00 03 00 0f 00 00 00     Channel: can0
Timestamp: 1729976462.338054    ID: 1001d085    X Rx                DL:  1    01                          Channel: can0
Timestamp: 1729976462.338326    ID: 0001d085    X Rx                DL:  0                                Channel: can0

**Reply by gemu2015** [reply]
i do not have any Sorel controller. i simply got this factor from the script referenced in a post above (https://github.com/ddimension/can-ltdc-logger/blob/master/can-ltdc-logger.c)

**Comment by javier6d2** [comment]
Summary:

Compile firmware with `#define USE_SML_CANBUS`. Use a proper CAN transceiver. You need a **120Ohms** resistor in the second CAN port of the controller, otherwise ESP32 device will not decode anything on the bus, the baudrate is **250 KBITS**.

If decoding works you should see a stream of CAN telegrams with regular updates at least every 20sec. 

Script:

```
>D

>B
=>sensor53 r

>M 1
; SOREL LTDC
; params -> 04 = baudrate 250kb + number of receive buffers * 100
+1,3,C,0,3204,CAN,1,5
1,100124800500uuUU@10,Temp S1,ºC,S1,0
1,100124800501uuUU@10,Temp S2,ºC,S2,0
1,100124800502uuUU@10,Temp S3,ºC,S3,0
1,100124800503uuUU@10,Temp S4,ºC,S4,0
1,10022480050000ss@-0.01,Relay R1,,R1,0
1,10022480050100ss@-0.01,Relay R2,,R2,0
1,10022480050200ss@-0.01,Relay R3,,R3,0
#
```

![imagen](https://github.com/arendst/Tasmota/assets/16863195/0a8bc021-cd19-4654-bf5c-d8dfc92ca917)

**Reply by WillySteenbergen** [reply]
Hi, Do you have the compiled firmware for this, and the wire diagram?
